[FSDP] Fixed init for 1-way FSDP with TP

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120244

For 1-way FSDP, we convert to `NO_SHARD`. This avoids doubling memory usage by keeping both a 'sharded' and 'unsharded' flat parameter of the original size.

However, for `NO_SHARD`, `use_sharded_views()` directs to `use_unsharded_views()` since we have the unsharded data. `use_unsharded_views()` for `DTensor` specifically re-wraps the `DTensor` view in `nn.Parameter` since we cannot do `.data` assignment. In that case, the new `nn.Parameter` object lost its attribute annotating it as FSDP-flattened, and the `DTensor` `nn.Parameter` will be incorrectly reconsidered for a parent FSDP module's flat parameter.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
[FSDP] Fixed init for 1-way FSDP with TP

[ghstack-poisoned]

mvpatel2000 commented:
@awgu seeing this patch the wrap error but fail on the full training run:
```
_ /usr/lib/python3/dist-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py:835 in      _
_ forward                                                                                          _
_                                                                                                  _
_    832 _   _   ):                                                                                _
_    833 _   _   _   args, kwargs = _root_pre_forward(self, self, args, kwargs)                    _
_    834 _   _   _   unused = None                                                                 _
_ _  835 _   _   _   args, kwargs = _pre_forward(                                                  _
_    836 _   _   _   _   self,                                                                     _
_    837 _   _   _   _   handle,                                                                   _
_    838 _   _   _   _   _pre_forward_unshard,                                                     _
_                                                                                                  _
_ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_runtime_utils.py:413 in _pre_forward      _
_                                                                                                  _
_    410 _   _   if handle:                                                                        _
_    411 _   _   _   handle._training_state = HandleTrainingState.FORWARD                          _
_    412 _   _   if unshard_fn is not None:                                                        _
_ _  413 _   _   _   unshard_fn(state, handle)                                                     _
_    414 _   _   # Register post-backward hooks to reshard the parameters and reduce-scatter       _
_    415 _   _   # their gradients. They must be re-registered every forward pass in case          _
_    416 _   _   # the `grad_fn` is mutated.                                                       _
_                                                                                                  _
_ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_runtime_utils.py:448 in                   _
_ _pre_forward_unshard                                                                             _
_                                                                                                  _
_    445 _   # If the handles have been prefetched, then there is no need to call                  _
_    446 _   # `_unshard()` again                                                                  _
_    447 _   if not handle._prefetched:                                                            _
_ _  448 _   _   _unshard(state, handle, state._unshard_stream, state._pre_unshard_stream)         _
_    449 _   handle._needs_pre_forward_unshard = False                                             _
_    450 _   # Don't wait during trace                                                             _
_    451 _   if not torch.distributed._functional_collectives.is_torchdynamo_compiling():          _
_                                                                                                  _
_ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_runtime_utils.py:321 in _unshard          _
_                                                                                                  _
_    318 _   if not handle:                                                                        _
_    319 _   _   return                                                                            _
_    320 _   with state._device_handle.stream(pre_unshard_stream):                                 _
_ _  321 _   _   ran_pre_unshard = handle.pre_unshard()                                            _
_    322 _   if ran_pre_unshard:                                                                   _
_    323 _   _   unshard_stream.wait_stream(pre_unshard_stream)                                    _
_    324 _   if state.limit_all_gathers:                                                           _
_                                                                                                  _
_ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_flat_param.py:1252 in pre_unshard         _
_                                                                                                  _
_   1249 _   _   _   self._use_sharded_views()                                                     _
_   1250 _   _   ret = False                                                                       _
_   1251 _   _   if self._use_orig_params and not self._skip_writeback_check:                      _
_ _ 1252 _   _   _   ret = self._writeback_orig_params()                                           _
_   1253 _   _   if (                                                                              _
_   1254 _   _   _   self.uses_sharded_strategy                                                    _
_   1255 _   _   _   and not self._offload_params                                                  _
_                                                                                                  _
_ /usr/lib/python3/dist-packages/torch/utils/_contextlib.py:115 in decorate_context                _
_                                                                                                  _
_   112 _   @functools.wraps(func)                                                                 _
_   113 _   def decorate_context(*args, **kwargs):                                                 _
_   114 _   _   with ctx_factory():                                                                _
_ _ 115 _   _   _   return func(*args, **kwargs)                                                   _
_   116 _                                                                                          _
_   117 _   return decorate_context                                                                _
_   118                                                                                            _
_                                                                                                  _
_ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_flat_param.py:2254 in                     _
_ _writeback_orig_params                                                                           _
_                                                                                                  _
_   2251 _   _   _   _   flat_param._params[i] = param                                             _
_   2252 _   _   _   if needs_param_writeback:                                                     _
_   2253 _   _   _   _   expected_shape = torch.Size([numel_in_shard])                             _
_ _ 2254 _   _   _   _   self._writeback_tensor(                                                   _
_   2255 _   _   _   _   _   param, flat_param, i, expected_shape, offset_in_shard, True           _
_   2256 _   _   _   _   )                                                                         _
_   2257 _   _   _   _   wroteback = True                                                          _
_                                                                                                  _
_ /usr/lib/python3/dist-packages/torch/distributed/fsdp/_flat_param.py:2353 in _writeback_tensor   _
_                                                                                                  _
_   2350 _   _   _   # NOTE: Gradient shape mismatch is not possible in practice since             _
_   2351 _   _   _   # the gradient shape is enforced to match that of the parameter and           _
_   2352 _   _   _   # we already check for parameter shape mismatch.                              _
_ _ 2353 _   _   _   raise RuntimeError(                                                           _
_   2354 _   _   _   _   f"Cannot writeback when the {'parameter' if is_param else 'gradient'} "   _
_   2355 _   _   _   _   f"shape changes\nExpects {expected_shape} but got {src_tensor.shape}"     _
_   2356 _   _   _   )                                                                             _
____________________________________________________________________________________________________
RuntimeError: Cannot writeback when the parameter shape changes
Expects torch.Size([132120576]) but got torch.Size([86016, 3072])
```

awgu commented:
@mvpatel2000 Sorry, I might need a repro again to dig into this :/

I am curious how valuable it is to apply 1-way FSDP. Is it mainly to avoid assigning those TP FFN parameters to the parent FSDP module, which would otherwise require using `ignored_modules`?

mvpatel2000 commented:
> @mvpatel2000 Sorry, I might need a repro again to dig into this :/

Working on it!

> I am curious how valuable it is to apply 1-way FSDP. Is it mainly to avoid assigning those TP FFN parameters to the parent FSDP module, which would otherwise require using `ignored_modules`?

If I understand correctly, you are suggesting the TP parallelism is applied to the entire model? This could be done if HSDP is applied to the other weights so they are reduced correctly, but it's a little hacky and annoying to switch between HSDP and FSDP.

awgu commented:
Sorry, to clarify: My understanding was that the issue is specific to N-way TP and 1-way FSDP on a module. In that case, the 1-way FSDP is not doing anything (i.e. no communication), so its only purpose is to be an annotation so that any parent FSDP modules to not take the 1-way FSDP module's parameters into their own flat parameter.

If that is the correct understanding, I was wondering if we can try to work around this issue, e.g. using `ignored_modules`. I am somewhat concerned that hardening `NO_SHARD` might be pretty involved (but we can see based on the issue).

mvpatel2000 commented:
@awgu I've added an updated minimal repro including a forward in the corresponding issue

> Sorry, to clarify: My understanding was that the issue is specific to N-way TP and 1-way FSDP on a module. In that case, the 1-way FSDP is not doing anything (i.e. no communication), so its only purpose is to be an annotation so that any parent FSDP modules to not take the 1-way FSDP module's parameters into their own flat parameter.
> 
> If that is the correct understanding, I was wondering if we can try to work around this issue, e.g. using `ignored_modules`. I am somewhat concerned that hardening `NO_SHARD` might be pretty involved (but we can see based on the issue).

If it's hard to fix FSDP for this case, I can try out this solution. Honestly, I'm not sure I completely understand the issue, so if you think it fixes the issue, happy to try it if the minimal repro doesn't work.

It's a little annoying to special case for this. I could add it to our core library so it's probably not too painful internally (beyond adding code for it), but I imagine it's unideal for users who aren't using our libraries and just torch.

Commit:
Update on "[FSDP] Fixed init for 1-way FSDP with TP"


For 1-way FSDP, we convert to `NO_SHARD`. This avoids doubling memory usage by keeping both a 'sharded' and 'unsharded' flat parameter of the original size.

However, for `NO_SHARD`, `use_sharded_views()` directs to `use_unsharded_views()` since we have the unsharded data. `use_unsharded_views()` for `DTensor` specifically re-wraps the `DTensor` view in `nn.Parameter` since we cannot do `.data` assignment. In that case, the new `nn.Parameter` object lost its attribute annotating it as FSDP-flattened, and the `DTensor` `nn.Parameter` will be incorrectly reconsidered for a parent FSDP module's flat parameter.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_flat_param.py:
```diff
@@ -2203,6 +2213,8 @@ def _writeback_orig_params(self) -> bool:
             if self.uses_sharded_strategy or not self._offload_params
             else flat_param._cpu_grad
         )
+        from torch.distributed._tensor import DTensor
```
I think we cannot import `DTensor` at top level or else it breaks internal torch packages or something.
